### ISSUE TITLE
Confusion on 'providedIn' for a specific NgModule

### ISSUE THREAD

**Issue by DanielYKPan** [angular/angular#25784]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[ X  ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
This is from the doc

`The @Injectable() decorator has the providedIn metadata option, where you can specify the provider of the decorated service class with the root injector, or with the injector for a specific NgModule.`

`Providing a service with the app's root injector is a typical case, and the CLI sets up this kind of a provider automatically for you when generating a new service. However, you might not always want to provide your service at the root level. You might, for instance, want users to explicitly opt-in to using the service.`

`Instead of specifying the root injector, you can set providedIn to a specific NgModule.`

`For example, in the following excerpt, the @Injectable() decorator configures a provider that is available in any injector that includes the HeroModule.`

What I understand from the doc, when you use `providedIn: 'root'` in your service, you don't have to configure this service at the app module level. And the service is available anywhere the class is imported. 

But if I use `providedIn: HeroModule`, I thought I didn't have to configure this service at the HeroModule level and I could use this service in HeroComponent (HeroComponent declared in HeroModule). This is not the case. And I still need to configure that service in HeroModule.

So What is this `providedIn: SpecificModule` for?

here is the [demo](https://stackblitz.com/edit/angular-1duwvx?file=src%2Fapp%2Fhero%2Fhero.service.ts)

## Environment

<pre><code>
Angular version: 6.*.*

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

**Comment by trotyl** [comment]
https://stackblitz.com/edit/angular-fcmdtf?file=src/app/hero/hero.service.ts

Please check the console logs, **you didn't configured it to `HeroModule`, but an `undefined` value**, so it's just ignored.

Duplicate of https://github.com/angular/angular/issues/24082.

**Comment by DanielYKPan** [comment]
Yes, I know it didn't work.
But I don't understand this `providedIn: SpecificModule` for.

The doc even have the example I mentions before.
```
import { Injectable } from '@angular/core';
import { HeroModule } from './hero.module';
import { HEROES } from './mock-heroes';

@Injectable({
  // we declare that this service should be created
  // by any injector that includes HeroModule.
  providedIn: HeroModule,
})
export class HeroService {
  getHeroes() { return HEROES; }
}
```

**Comment by trotyl** [comment]
It works as expected unless you're making circular dependencies between files.

The doc is correct, but also make sense if you're requesting some warning in doc.

**Comment by DanielYKPan** [comment]
If I declare the `HeroService` in `HeroModule`, and use `providedIn: HeroModule`, there would be circular dependencies warning. So How do I avoid that?

**Comment by trotyl** [comment]
Just don’t do that.

**Comment by DanielYKPan** [comment]
I know. If I don't do that, and only declare the HereService in Her0Module, everything works perfectly.
So back to the start point.
If I don't do that, 
what's the `providedIn: SpecificModule` for?

**Comment by trotyl** [comment]
> what's the `providedIn: SpecificModule` for?

`providedIn: SpecificModule` would guarantee that the service can only be used when `SpecificModule` being imported, same as configuring provider in its metadata.

**Comment by DanielYKPan** [comment]
From what I understand from @trotyl comment

`providedIn: SpecificModule would guarantee that the service can only be used when SpecificModule being imported, the same as configuring provider in its metadata.`

If I use  `providedIn: HeroModule` in `HeroService` (not declare in HeroModule) and import `HeroModule` in `AppModule`, I could inject the service in `AppComponent`.  But because `HeroService` not declare in `HeroModule`, `HeroComponent` could not inject `HeroService`.

**Comment by mlc-mlapis** [comment]
@DanielYKPan ... hmm, so the case when you want to instantiate such a service only on `HeroModule` level and have the possibility to tree-shake it if it is not injected by any part of `HeroModule` is still out of the game?

**Comment by DanielYKPan** [comment]
I guess so.

The doc is so confusing. The doc has the `HeroModule` and `HeroService` example codes, but never explain any use case. It mentions the benefit of `providedIn`, but never give any cases.

At the beginning,
The doc has `providedIn: 'root'`, and says the service could be used anywhere in the apps. So I thought if you use `provideIn: SpecificModule` (not declare that service in SpecificModule), you could only use that service in any components/directives/pipes in that module. But this is not the case.
So that means if you use `providedIn: SpecificModule`, you still need to declare that service in SpecificModule. But this would cause circular dependency warnings. (SpecificService imports SpecificModule file, and SpecificModule imports SpecificService file).

**Comment by trotyl** [comment]
@mlc-mlapis The `providedIn: MyModule` differs to `providedIn: 'root'` in two facts:

+ Require `MyModule` be available in current module transitive scope;
+ Register the provider to local injector which `MyModule` belongs to;

### Use case A: `NgModule` based configuration

If some service depends on other `InjectionToken`s, and these tokens were configured via `NgModule` static methods:

```typescript
@NgModule()
class MyModule {
  static config(params: SomeType) {}
}

@Injectable({
  providedIn: MyModule
})
class MyService {
  constructor(@Inject(SomeToken) params: SomeType) {}
}
```

But this is not the only way for guarding, one can also check for `SomeToken` and produce a better error message.

### Use case B: consume lazy-module providers

If `MyModule` is imported by some lazy-loaded module, then `MyService` is provided in that module injector, rather than application (root-module) injector.

So one can provide then deps in that lazy module, while `providedIn: 'root'` cannot retrieve it and results to error:

```typescript
@NgModule()
class MyModule { }

@NgModule({
  imports: [MyModule],
  providers: [
    { provide: SomeToken, useValue: someValue }
  ]
})
class LazyModule { }

@Injectable({
  providedIn: MyModule
})
class MyService {
  constructor(@Inject(SomeToken) params: SomeType) {}
}
```

But this could be some kind of fragile once `MyModule` being imported by root modules mistakenly.

---

There's no condition that definitely requiring `providedIn: MyModule`, but one could use that to make things easy sometimes.

**Comment by mlc-mlapis** [comment]
@trotyl ... thank you for showing the above use-cases which help things to be cleaner. 🎁

**Comment by stas-kh** [comment]
Guys, then I have a question regarding `Core Module`. Do we still need this kind of module in Angular 6 applications?

Because according to [Core Feature Module](https://angular.io/guide/styleguide#core-feature-module) this kind of module is used to: 
`Putting a singleton service whose instance will be shared throughout the application in the CoreModule (e.g. ExceptionService and LoggerService).`

For example. in my appliacation (written on Angular 5) the `core module` looked like:
```
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';

import { AuthGuardService } from './auth-guard/auth-guard.service';
import { AuthService } from './api/auth/auth.service';
import { MouseObserverService } from './mouse-observer/mouse-observer.service';
...

@NgModule({
  imports: [
    CommonModule,
    HttpClientModule
  ],
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: RootInterceptor,
      multi: true
    },
    AuthGuardService,
    AuthService,
    MouseObserverService.
    ...
  ]
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error('CoreModule is already loaded. Import it in the AppModule only');
    }
  }
}
```

But then I migrated the app to Angular 6 and set for those 3 modules `providedIn: 'root'` rule, so then they are available everywhere in the app. And seems this works fine.

Therefore, what is the intent of having `core module`? Just for the case if we need to provide some another service for the token (as it's done for `HTTP_INTERCEPTORS`)?

**Comment by mlc-mlapis** [comment]
@stas-kh ... probably don't understand. The `core` module (`main`, `app` ... or what ever is called ... the name is not important in fact) ... is the module which the module on bootstrap, so like:
JIT: `platformBrowserDynamic().bootstrapModule(AppModule);`
AOT: `platformBrowser().bootstrapModuleFactory(AppModuleNgFactory)`

**Comment by stas-kh** [comment]
@mlc-mlapis I mean the documentation recommends us to put all shared services into one module, called `core` module. This [Core Module](https://angular.io/guide/styleguide#core-feature-module)  must be imported only once and most probably into `AppModule`.

And for me, it looked reasonable to organize the application this way. But now there is an option to set `providedIn: 'root'` and then you don't have to put your services either to `AppModule` or to `CoreModule`.

And my question: Is there any profit of having this `Core Module` described in Style Guide? Or in the new applications written in Angular 6 it doesn't make any sense and we can avoid it

**Comment by mlc-mlapis** [comment]
@stas-kh ... I think that it's just the style ... with no real effect on functionality ... you can use it not ... it's on you.

**Comment by stas-kh** [comment]
@mlc-mlapis Yes, you are right.

But I'd like to hear an opinion of others.

**Comment by maxkorz** [comment]
@trotyl sorry, it's still confusing for me, could you please take a look at this stackblitz? https://stackblitz.com/edit/angular-lazy-load-o3zayc?file=app%2Fabout%2Fabout.service.ts

What's wrong with my code? 
```ts
@Injectable({ providedIn: AboutModule })
export class AboutService {
  constructor() {
    console.log('About Service constructor');
  }
}
```
AboutModule is a lazy-loading module:
```
export const AppRoutingModule: ModuleWithProviders = RouterModule.forRoot([
  { path: '', redirectTo: '/home', pathMatch: 'full' },
  { path: 'about', loadChildren: './about/about.module#AboutModule' }
]);
```

When I try to navigate to 'about' route, I get error:
```
ERROR Error: Uncaught (in promise): Error: StaticInjectorError(AppModule)[AboutComponent -> AboutService]: 
  StaticInjectorError(Platform: core)[AboutComponent -> AboutService]: 
    NullInjectorError: No provider for AboutService!
Error: StaticInjectorError(AppModule)[AboutComponent -> AboutService]: 
  StaticInjectorError(Platform: core)[AboutComponent -> AboutService]: 
    NullInjectorError: No provider for AboutService!
```

**Comment by trotyl** [comment]
@maxkorz Why do you think `providedIn: undefined` could work? As clarified in both issues you cannot make circular dependencies between files. It's not an Angular BUG as you just passed `undefined` to `providedIn`, it's not designed to work anyway.

How to prevent circular dependencies is your design problem, you can either stop self-consuming or create a new `AboutProvidersModule` internal NgModule for service registry.

**Comment by maxkorz** [comment]
@trotyl 
> Why do you think providedIn: undefined could work

because:
1. Nothing in the documentation indicates that `provdedIn: FeatureModule` won't work with lazy-loading modules.
2. It's not obvious that `AboutModule` is `undefined` when you do `import { AboutModule } from './about.module';` in a lazy-loading code.
3. Documentation encourages you to use `provdedIn: FeatureModule`:
https://angular.io/guide/hierarchical-dependency-injection#injectable-level-configuration
![image](https://user-images.githubusercontent.com/25089646/45032282-22ab5880-b05a-11e8-9e85-6f4c2073c284.png)

> It's not an Angular BUG

I think it's a documentation issue.

**Comment by manklu** [comment]
@maxkorz 
> What's wrong with my code?

You need an additional module

https://stackblitz.com/edit/angular-lazy-load-jxjhfw?file=app%2Fabout%2Fabout.service.ts

**Comment by jenniferfell** [comment]
Interesting thread. I could use some help determining possible next actions (if any): 
What doc is quoted above?
What other doc did you use?
Do you have a specific recommendation?

Thanks.

**Comment by GeorgeKnap** [comment]
> @maxkorz
> 
> > What's wrong with my code?
> 
> You need an additional module
> 
> https://stackblitz.com/edit/angular-lazy-load-jxjhfw?file=app%2Fabout%2Fabout.service.ts

I think this method should be put into Angular docs as I was facing same issue as all others.

**Comment by bennadel** [comment]
I've been battling this issue for the past 3-hours. What made it super confusing for me was that I have two services in a module that both use `providedIn: MyModule` syntax. The **first one will work** (and can be injected into a Component. But, the **second one breaks** with the `NullInjectorError: No provider` error. Despite the fact that they are using the exact same syntax and file paths.

That is extremely confusing! And the idea that there is a "circular reference" doesn't hold much water in my case because it would / should be breaking for both Service, not just the second one. It's like the meta-data gets nullified after the first.

**Comment by kapunahelewong** [comment]
@stas-kh to answer this question you asked:

> Guys, then I have a question regarding Core Module. Do we still need this kind of module in Angular 6 applications?

According to @DeborahK's [insight](https://github.com/angular/angular/issues/17825#issuecomment-435535268), no, we don't so, I'm currently working on a PR to remove it. 🎉I've got it out of the copy (written explanations) and am refactoring the NgModules example, which still uses it. 

Thanks for a great question.

**Comment by stas-kh** [comment]
@kapunahelewong Thank you. Looking forward to getting the docs updated

**Comment by kidqn** [comment]
provdedIn: FeatureModule cannot work with lazy load Module! update docs please :D

**Comment by kapunahelewong** [comment]
@DanielYKPan This is a work in progress, so some of the details will change, but does this help answer your concern? [#28700](https://github.com/angular/angular/blob/6b5597ee72dcab09026670db3c15e726659c7c7a/aio/content/guide/hierarchical-dependency-injection.md#providedin-a-feature-module)

If so, I'd like to add the `state: has PR label`.

**Comment by kapunahelewong** [comment]
This issue might also be of interest: https://github.com/angular/angular/issues/29038

And this repo shows how `ModuleInjector` works : https://github.com/kapunahelewong/module-injector-tree

They [Hierarchical Injector guide](https://angular.io/guide/hierarchical-dependency-injection) has also been extensively rewritten. Hope this helps!

**Comment by Elrashid** [comment]
I'm not sure if this will help someone , 
but i got 
***Circular dependency detected*** 
and 
***Cannot access 'XxModule' before initialization**** 
when changeing 
***providedIn: 'root'***
to 
***providedIn: XxModule*** ,

so I import Injectable Service manually using providers array 

Ref : [angular - error: "component factory found for" when using dialog.open form angluer service even though it's included in entryComponents - Stack Overflow](https://stackoverflow.com/q/57967381/4879683)

**Comment by kapunahelewong** [comment]
Thanks @Elrashid! Do you mean the `@Component()` `providers` array? Because that will indeed work, but now the service is visible in a different injector tree (the `ElementInjector`). In that case, the service will be visible to that component and its children and this visibility is modifiable with `@Host()`, `@SkipSelf()` etc.

Maybe you mean the `@NgModule()` `providers` array. In which case, this works too, but isn't tree shakable. (This way is in the `ModuleInjector`, not the `ElementInjector`).

To have your service be tree shakable and use lazy loading, take a look at [this example](https://github.com/kapunahelewong/module-injector-tree/tree/master/src/app):
1. In the service, use `providedIn` with a certain NgModule (step 2 below): https://github.com/kapunahelewong/module-injector-tree/blob/master/src/app/animal/animal.service.ts#L5
1. Give the service its own module: That's the `AnimalModule` the `AnimalService` is provided in.
1. In your lazy loaded module, import the service's module (`AnimalModule`) https://github.com/kapunahelewong/module-injector-tree/blob/master/src/app/lazy/lazy.module.ts

**Comment by cpyle0819** [comment]
> This issue might also be of interest: #29038
> 
> And this repo shows how `ModuleInjector` works : https://github.com/kapunahelewong/module-injector-tree
> 
> The [Hierarchical Injector guide](https://angular.io/guide/hierarchical-dependency-injection) has also been extensively rewritten. Hope this helps!

@kapunahelewong you're a hero. Thank you for this!

**Comment by kapunahelewong** [comment]
Thank you, @cpyle0819! 😊If you see room for improvement in the `ModuleInjector` demo or doc, just let me know. 🌺

**Comment by mcalmus** [comment]
The way I've solved the circular dependencies issue is to create a separate service module within the same context as my main component module. This is similar to what some documentation has shown, but I think most are looking for a module-specific service and aren't thinking to have it in an outside module. So:

src/app/my-module:
    my-module.module.ts
    my-module.component.ts
    my-service-module.module.ts
    my-service-module.service.ts

You can then have my-module import my-service-module and have my-service-module.service be provied in my-service-module.

**Comment by rockument69** [comment]
@kapunahelewong : I see in the comments that "The [Hierarchical Injector guide](https://angular.io/guide/hierarchical-dependency-injection) has also been extensively rewritten." Can I now close this? Thanks.

**Comment by kapunahelewong** [comment]
@rockument69, I don't think we should close this one yet because as far as I know, none of the DI documentation includes a demo of `providedIn: someModule`. I think this issue warrants its own new doc and focused example, tbh. The rewrite of the [Hierarchical Injectors guide](https://angular.io/guide/hierarchical-dependency-injection) helps in that it shows the two injectors and how they differ, and has a minor mention about being able to provide somewhere other than root, but it doesn't demo or explain it. And extrapolating the technique from the current docs isn't obvious.

@DanielYKPan and others in this thread, please feel free to chime in here. What do you all think?

**Comment by Blackbaud-SteveBrush** [comment]
Is it okay to use `providedIn: 'any'` for services that are part of a lazy-loaded module? Seemed to work for me:
```
@Injectable({
  providedIn: 'any'
})
export class LazyService {
  constructor() {}
}
```
Docs: https://angular.io/guide/providers#limiting-provider-scope-by-lazy-loading-modules

**Comment by yarinsa** [comment]
Hey There!
I have been battling with the same question.
I am running an electron app that runs separate windows and each window initiate other module. So I need some services to be initiated only within a module scope.
I am getting 'Uncaught ReferenceError: Cannot access '${name}Module' before initialization'.
Did any of you figure out what is the best way to do it?
I was able to provide it within the module itself, but will be much cleaner if each service will have it in it's metadata. And then the co-dependent services doesn't work.
`    { provide: CommandBarService, deps: [BarService, PopupService] }`

**Comment by antoantonydh** [comment]
I tried like this and it is working.

```
@Injectable({ providedIn: forwardRef(() => AboutModule)})
export class AboutService {
  constructor() {
    console.log('About Service constructor');
  }
}
```
lazy-loading About module:

```
export const AppRoutingModule: ModuleWithProviders = RouterModule.forRoot([
  { path: '', redirectTo: '/home', pathMatch: 'full' },
  { path: 'about', loadChildren: './about/about.module#AboutModule' }
]);
```

**Comment by amitbeck** [comment]
> I have the same problem in Angular 13: Uncaught (in promise): ReferenceError: Cannot access 'ContactsModule' before initialization
> 
> I am layz loading modules

@gregoriusus See https://github.com/angular/angular/issues/25784#issuecomment-592565166, worked for me

**Comment by daziz-quotient** [comment]
Hi @JeanMeche. 

> The replacement is to use `providedIn: 'root'`.

The replacement you suggested in the closing comment for this issue doesn't seem a suitable replacement if someone is looking for a replacement in this thread.

The discussion started in the first place to find out what's going wrong when using `{ providedIn: someLazyModule }` in a service belonging to a lazily loaded module. And this action was being performed to basically limit the use of a service in `someLazyModule` unless some other module imports the `someLazyModule`. Going through this whole thread today and experimenting, I believe below are the 2 replacements:

1. Now that `providedIn` is deprecated, I'd suggest not to use it anymore. And a direct replacement would be to provide the service in the `providers` array of the `someLazyModule`
2. The only use for `providedIn` seems to be for the enabling of tree-shaking when that service isn't used anywhere. So for people who are still planning to use `providedIn` for tree-shaking feature. You need to follow answer give [here](https://github.com/angular/angular/issues/25784#issuecomment-532337542).

> The solution that the answer in point 2 is basically. If we want to use providedIn and enable tree shaking, we need to create a dedicated module for that service and import that service in a lazy module. The service module will have only the service, it cannot have a component inside that tries to use the service, because that creates a circular dependency and this is what initiated this thread in the first place


@kapunahelewong , if you or someone from the contributing team can give a sign-off to my findings and suggestions here or suggest an edit if required, that would be really helpful for me and even for future stumblers

**Comment by daziz-quotient** [comment]
Also, is there anything existing in Angular or is there something in the pipeline which can give devs the ability write code in such a way that will enable effective tree shaking for services and lazy modules? As seeing the documents, at least 1 of the main intended use of `providedIn: someModule` was to enable effective tree-shaking.
@kapunahelewong @JeanMeche

**Comment by JeanMeche** [comment]
`providedIn: 'root'` is what you are looking for.

**Comment by daziz-quotient** [comment]
Umm, I'm sorry, but this confuses me. According to docs, `providedIn: 'root'` registers the service in the root injector context. And any module's components can access it by directly importing it

How can one scope a service to its own lazy-loaded module(which was what this thread was about inherently)?

Is there a new documentation available which explains how `providedIn: 'root'` can achieve this?

**Comment by daziz-quotient** [comment]
Thank you 👍
